### PR TITLE
Dynamic transpose kernels

### PR DESCRIPTION
**Pull request by zjing14**
- Convert transpose kernels in MIOpenUtilKernels4.cl to dynamic kernels (Issue #482)

### CODE REVIEW

**Comment by zjing14** [comment]
@JehandadKhan @daniellowell Updates: 
1) Experiments to identify the cause of failures in test_tensor_trans. First, I just keep a single kernel in the kernel file, still got failure. So, the failure is not due to call the wrong kernel. Second, if I ran the same multiple transposes individually with the same cache kernels, no issue. Thus, I think there is something wrong with the host side or test program.
2) The slow down problem is due to fluctuated performance of rocblas gemm with dynamic transpose. The performance is stable in the develop branch. But with dynamic transpose, the rocblas gemm performance drop significantly in some runs  That is weird.

File: src/ocl/utilocl.cpp [hunk 1]
```diff
@@ -699,107 +699,119 @@ float transpose_NCHW2CNHW(const Handle& handle,
...
+        }
 
+        kernel_name += "_" + READ_TYPE;
+
+        auto&& kernels = handle.GetKernels(kernel_name, network_config);
+        if(!kernels.empty())
+        {
+            auto kernel = kernels.front();
+            kernel.ldims = {vld[0], vld[1], vld[2]};
+            kernel.gdims = {vgd[0], vgd[1], vgd[2]};
```
Comment (JehandadKhan) [inline src/ocl/utilocl.cpp:747]: @zjing14 Here is the fix to the issue you were seeing.

Comment (zjing14) [reply src/ocl/utilocl.cpp]: I see. Thanks.

**Comment by daniellowell** [comment]
Coding errors:
http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/dynamic_UtilKernels4/9/pipeline

**Comment by JehandadKhan** [comment]
@zjing14 Test failure in convolution

File: src/ocl/convolutionocl.cpp [hunk 2]
```diff
@@ -87,24 +87,9 @@ static inline bool IsAnyBufferBF16(const TensorDescriptor& xDesc,
...
 
-size_t GetKernelGlobalWorkDim(const KernelInvoke& kernel, int dim)
-{
-#if(MIOPEN_BACKEND_HIP)
-    return kernel.gdims[dim];
-#else
-    return kernel.global_work_dim[dim];
-#endif
-}
+size_t GetKernelGlobalWorkDim(const KernelInvoke& kernel, int dim) { return kernel.gdims[dim]; }
```
Comment (daniellowell) [inline src/ocl/convolutionocl.cpp:90]: What is this change for?

Comment (zjing14) [reply src/ocl/convolutionocl.cpp:90]: Changed it for updating gdims for each run, correct? @JehandadKhan

Comment (JehandadKhan) [reply src/ocl/convolutionocl.cpp:90]: That is correct, I updated the member names to make them uniform between both the HIP and the OpenCL backend

Comment (daniellowell) [reply src/ocl/convolutionocl.cpp:90]: Curious that they were different in the first place. Are you sure there are no side effects (usages of the old style)?

**Comment by JehandadKhan** [comment]
@zjing14 can you please fix the CI issues so we can get this in ?

**Comment by zjing14** [comment]
> @zjing14 can you please fix the CI issues so we can get this in ?

The CI is still running without an issue, though it has red "X".

**Comment by zjing14** [comment]
CI passed.

**Review by JehandadKhan — APPROVED** [review]

**Pull request merged**